### ISSUE TITLE
[clang] False-positive with -Wuninitialized or not apply infered value to branch?

### ISSUE THREAD

**Issue by hstk30-hw** [llvm/llvm-project#85372]
https://godbolt.org/z/Efxa6Yoca

```c
extern void use(int);

void test() {
    int v;
    _Bool s = 1;

    if (sizeof(int) == 1)
        v = 3;
    else 
        s = 0;  // clang and gcc know that will only reach here
    
    if (s) {   // but clang not apply  `s=0`  to here
        use(v);
    }
}
```

And clang report:

```
<source>:13:13: warning: variable 'v' is uninitialized when used here [-Wuninitialized]
        use(v);
            ^
<source>:4:10: note: initialize the variable 'v' to silence this warning
    int v;
         ^
          = 0
```

**Comment by hstk30-hw** [comment]
CC @aaronpuchert @shafik @AaronBallman

**Comment by shafik** [comment]
Confirmed but not sure if this is known deficiency or not.

**Comment by aaronpuchert** [comment]
The Clang CFG marks `v = 3` as unreachable because the condition is a compile-time constant. However, `s` is not compile-time constant. What happens before that is irrelevant. So we can even simplify this to:
```c++
void test() {
    int v;
    _Bool s = 0;
 
    if (s) {
        use(v);
    }
}
```
and still get a warning. But exploring different execution paths or collecting all possible values of a variable is outside of the scope for the core compiler. That is the domain of the static analyzer, in this case [core.CallAndMessage](https://clang.llvm.org/docs/analyzer/checkers.html#core-callandmessage-c-c-objc). This is on by default, so you can simply run Clang with `--analyze`, but you get better output if you add `-Xanalyzer -analyzer-output=text`. Your example produces no warnings, but when I swap `s = 0` with `s = 1`:
```
<stdin>:13:9: warning: 1st function call argument is an uninitialized value [core.CallAndMessage]
   13 |         use(v);
      |         ^   ~
<stdin>:4:5: note: 'v' declared without an initial value
    4 |     int v;
      |     ^~~~~
<stdin>:7:5: note: Taking false branch
    7 |     if (sizeof(int) == 1)
      |     ^
<stdin>:12:9: note: 's' is true
   12 |     if (s) {   // but clang not apply  `s=0`  to here
      |         ^
<stdin>:12:5: note: Taking true branch
   12 |     if (s) {   // but clang not apply  `s=0`  to here
      |     ^
<stdin>:13:9: note: 1st function call argument is an uninitialized value
   13 |         use(v);
      |         ^   ~
```
So I think `-Wuninitialized` works as designed, and if you want something more sophisticated, you need to use the Clang static analyzer.

**Comment by aaronpuchert** [comment]
Maybe a bit more detail why I think this is out of scope of the core compiler. The example is not typical. If you have obviously unreachable code, you'd just remove it, and if you have branches that are always taken or always not taken, you'd probably drop them or at least make the condition a compile-time constant.

So the branches that remain in a typical program are probably sometimes taken and sometimes not, depending on runtime circumstances. In those cases, reasoning like one might do for the given example doesn't get us very far. We have a couple of bad options:
* Explore execution paths. Ideally we'd explore all of them, but this isn't really feasible. So we have to stop exploration after some number of steps and work with what we've got. This makes the analysis a bit unpredictable. But it's still valuable, hence why it is available as a separate tool (the analyzer).
* Try to derive invariants that hold on join points. This could at best be done in a heuristic fashion, and even then it's more of a research topic.

That's why I think the stricter rule imposed by `-Wuninitialized` is Ok: it checks that every load is preceded by a store, regardless of which paths are actually feasible, pruning only branches when their condition is a compile-time constant according to language rules.

**Comment by aaronpuchert** [comment]
Oh, and to silence `-Wuninitialized`, just turn `s` into a proper compile-time constant:
```c
extern void use(int);

void test() {
    int v;
    const _Bool s = (sizeof(int) == 1) ? 1 : 0;

    if (sizeof(int) == 1)
        v = 3;
    
    if (s) {
        use(v);
    }
}
```

**Comment by hstk30-hw** [comment]
Thx your patient reply @aaronpuchert . 

> The example is not typical. If you have obviously unreachable code, you'd just remove it, and if you have branches that are always taken or always not taken, you'd probably drop them or at least make the condition a compile-time constant.

This example is simplified from a big  complicate project. The real code is mixed many macros, so the `uninitialized value` warning is hiddened deeply, recognize it is hard. That's why we need a smart code or sema checker.

To my surprise，GCC can recognize this situation. I will check gcc's implement.

**Comment by aaronpuchert** [comment]
GCC has [warnings in the middle-end](https://developers.redhat.com/blog/2019/03/13/understanding-gcc-warnings-part-2), and their `-Wuninitialized` is one of them. Clang deliberately doesn't do that. If you want behavior similar to GCC, you'll have to go with the static analyzer check.

**Comment by aaronpuchert** [comment]
By the way: with the first branch not compile-time constant, `v` would be considered conditionally initialized and the warning wouldn't fire. But since the first branch is decided at compile time, we see no write to `v`, so it counts as unconditionally uninitialized. So another way to fix the warning is to turn the branch condition into something that's not compile-time constant, for example by routing it through a non-const variable. This gives you two options:
* make sure that both branches can be decided at compile time,
* make sure that both branches can't be decided at compile time.

We might let the warning look into unreachable basic blocks, but this could break something else. And to my knowledge, all other flow-based warnings skip unreachable blocks, too.

**Comment by Kojoley** [comment]
> GCC has [warnings in the middle-end](https://developers.redhat.com/blog/2019/03/13/understanding-gcc-warnings-part-2), and their `-Wuninitialized` is one of them. Clang deliberately doesn't do that. If you want behavior similar to GCC, you'll have to go with the static analyzer check.

GCCs `-Wuninitialized` is one of the most hated warnings. High volume of false positives due to reliance on middle-end optimizations to suppress them made its reputation so bad that it's now turned off in [so many projects](https://github.com/search?q=%22-Wno-uninitialized%22&type=code), and in the end GCC moved non-control-flow aware warning to `-Wmaybe-uninitialized` (which is annoyingly turned on by `-Wuninitialized`/`-Wall`).

**Comment by AaronBallman** [comment]
The explanation by @aaronpuchert is correct; this is behaving by design and the static analyzer needs to be invoked for more precise analysis -- closing the issue as we explicitly do not want to follow GCC's behavior in this case.
